pyccel/parser/semantic.py

```diff
@@ -1682,6 +1682,7 @@ def _ensure_inferred_type_matches_existing(self, class_type, d_var, var, is_auga
                         status = 'unallocated'
 
                     new_expressions.append(Allocate(var, shape=d_var['shape'], status=status))
+                    self._allocs[-1].add(var)
```

Comment:
This is the wrong place for this line. The variable should only be added to `_allocs` if it's new (i.e. if the status is `unallocated`). The set should take care of this but it's better to be on the safe side

Reply:
so the `else` block where `status` is assigned `unallocated`.
may I know what does this line do and what's `new_expression` used for:
```python
new_expressions.append(Allocate(var, shape=d_var['shape'], status=status))
```

Reply:
Sometimes while creating nodes you need to create more nodes than expected. For example when translating `a = np.ones(4)` you need 2 nodes not one. The nodes are `[Allocate(a, 4), Assign(a, 1)`. Or when calling a function with a vector expression e.g. `f(a+b)` you may need to create a temporary to store the argument in. The list `new_expression` is designed to hold these unexpected nodes that are created in the functions `_ensure_inferred_type_matches_existing` or `_assign_lhs_variable` neither of which are supposed to create nodes usually.

This line creates the necessary `Allocate` node to allocate the data. The purpose of the `Allocate` node is described in its docstring:
https://github.com/pyccel/pyccel/blob/5a75f34563bd1ea3287afc29949bca98400be6fe/pyccel/ast/core.py#L371-L399

Reply:
check bacd8b3a and let me know if it's alright

Reply:
Thanks for the explanation!

Reply:
This looks fine but could be better. For execution speed it is always better to avoid `if` conditions. Here your new `elif` is unnecessary. You could just use `else`  or put the line in a slightly different location where the value has already been checked

Reply:
I first wrote: 
```python
 else:
                        status = 'unallocated'
                        self._allocs[-1].add(var)
```
but didn't like how it looked so for more clarity i added the `elif` condition, but I understand the execution speed concerns

Reply:
> but didn't like how it looked so for more clarity

That's fine. But why:
```python
                    if status != 'unallocated':
                        errors.report(ARRAY_REALLOCATION, symbol=var.name, ...)
                    elif status == 'unallocated':
                        self._allocs[-1].add(var)
```
instead of:
```python
                    if status != 'unallocated':
                        errors.report(ARRAY_REALLOCATION, symbol=var.name, ...)
                    else:
                        self._allocs[-1].add(var)
```

Reply:
Or:
```python
                    if status == 'unallocated':
                        self._allocs[-1].add(var)
                    else:
                        errors.report(ARRAY_REALLOCATION, symbol=var.name, ...)
```

Reply:
Unknown is for situations like:
```python
def f(b : bool):
   a : 'int[:]'
   if b:
       a = np.ones(4)

    a = np.zeros(4) # Does "a" need allocating or does it already exist with the correct size?
```

Reply:
But as you say `'unknown' != 'unallocated'` so this shouldn't affect whether an `if` is needed here or not

Reply:
> Or:
> 
> ```python
>                     if status == 'unallocated':
>                         self._allocs[-1].add(var)
>                     else:
>                         errors.report(ARRAY_REALLOCATION, symbol=var.name, ...)
> ```

this one better.

Reply:
updated 4e91d41b